### PR TITLE
18MS: Implement Meridian & Memphis Railway private

### PR DESCRIPTION
**Pull request by perwestling**
This private has a one time ability (for a corporation) to place
a token at half price. A similar way of specifying this as for
train_discount is used. If the value is between 0 and 1, it is
how big part is a discount of the tokening price. So 0.5 means
half price.

Use similar method as was done in 18GA Waycross & Southern to
just use ability if ability is actived explicitly.

Have made a change to hex selection. If hexes is empty this
means all (connected) hexes is allowed.

Use of ability does not close the private.

### CODE REVIEW

**Comment by perwestling** [comment]
The method used to have hexes as [] in the token ability, and use that as any-hex might not be the correct solution. Compare tile_lay for NWR in 1882.

**Comment by perwestling** [comment]
I suspect the implementation might be wrong, allowing for teleportation of token. So I need to try this further.

**Comment by perwestling** [comment]
If one do attempt to token a non-connected city, you get the game error "Cannot place token on XX because it is not connected".
I think this can be good enough.

File: lib/engine/config/game/g_18_ms.rb [hunk 1]
```diff
@@ -179,7 +179,17 @@ module G18MS
...
          "revenue":15,
          "desc":"The owning Major Company may lay their cheapest available token for half price. This is not an extra token placement. This ability can only be used once, and using it does not close the private company.",
-         "sym":"M&M"
+         "sym":"M&M",
+         "abilities": [
+            {
+              "type": "token",
+              "owner_type":"corporation",
+              "hexes": [],
+              "price": 0.5,
```
Comment (tobymao) [inline lib/engine/config/game/g_18_ms.rb:188]: would it be better to call it a discount? what do the other things do?

**Comment by tobymao** [comment]
still has unaddressed comments and merge conflict

File: lib/engine/config/game/g_18_ms.rb (hunk 1 above)
Comment (perwestling) [reply lib/engine/config/game/g_18_ms.rb]: I have now added a new attribute 'discount' that is used for this purpose.

**Comment by perwestling** [comment]
Have now updated the commit, and rebased towards latest master.

File: lib/engine/ability/token.rb [hunk 2]
```diff
@@ -5,14 +5,15 @@
 module Engine
   module Ability
     class Token < Base
-      attr_reader :hexes, :price, :teleport_price, :extra, :from_owner
+      attr_reader :hexes, :price, :teleport_price, :extra, :from_owner, :discount
 
-      def setup(hexes:, price:, teleport_price: nil, extra: nil, from_owner: nil)
+      def setup(hexes:, price: nil, teleport_price: nil, extra: nil, from_owner: nil, discount: nil)
```
Comment (tobymao) [inline lib/engine/ability/token.rb:10]: why did you change price?

File: lib/engine/step/tokener.rb [hunk 3]
```diff
@@ -72,17 +72,25 @@ def adjust_token_price_ability!(entity, token, hex)
...
           return [token, ability]
         end
 
         [token, nil]
       end
+
+      private
+
+      def new_token_price(token, ability)
+        return ability.price if ability.price
```
Comment (tobymao) [inline lib/engine/step/tokener.rb:90]: shouldn't this just be abstracted in the ability class?

in class ability, you can have the price reflect the discount

Comment (perwestling) [reply lib/engine/step/tokener.rb]: Have now replaced price attribute with price method in ability.

File: lib/engine/ability/token.rb (hunk 2 above)
Comment (perwestling) [reply lib/engine/ability/token.rb:10]: In case if discount used there is no meaning of price, so there is no reasonable default value.
Should I add a guard/game_error in case both price and discount are nil?
